Fixture redirects

FloEdelmann opened the pull request:
Closes #483, see also https://github.com/OpenLightingProject/open-fixture-library/pull/460#issuecomment-382099997

For an example, go to https://open-fixture-library-pr-484.herokuapp.com/minuit-une/m-carre

FloEdelmann commented:
<!-- GITHUB-TEST: tests/github/export-diff.js -->
# Plugin export diff
(Output of test script `tests/github/export-diff.js`.)

You can run view your uncommited changes in plugin exports manually by executing:
`$ node cli/diff-plugin-outputs.js -p <plugin name> <fixtures>`

## Fixture `minuit-une/m-carre` modified in this PR
Fixture output to all plugins is checked.

### :warning: Plugin `d-light`
*Removed files*
- minuit-une/m-carre-25ch.xml
- minuit-une/m-carre-40ch.xml

*Added files*
- minuit-une/ivl-carre-25ch.xml
- minuit-une/ivl-carre-40ch.xml


### :warning: Plugin `ecue`
<details>
<summary><b>Changed outputted file <code>UserLibrary.xml</code></b></summary>

```diff
--- removed
+++ added
@@ -2,9 +2,9 @@
 <Document Owner="user" TypeVersion="2" SaveTimeStamp="2018-04-20#14:59:56">
     <Library>
         <Fixtures>
             <Manufacturer _CreationDate="2018-04-20#14:59:56" _ModifiedDate="2018-04-20#14:59:56" Name="Minuit Une" Comment="" Web="https://minuitune.com/">
-                <Fixture _CreationDate="2018-03-30#00:00:00" _ModifiedDate="2018-03-30#00:00:00" Name="M-Carré (25ch mode)" NameShort="mcarre-25ch" Comment="generated by the Open Fixture Library – https://open-fixture-library.org/minuit-une/m-carre" AllocateDmxChannels="25" Weight="13" Power="125" DimWidth="594.3" DimHeight="200" DimDepth="551.9">
+                <Fixture _CreationDate="2018-03-30#00:00:00" _ModifiedDate="2018-04-20#00:00:00" Name="IVL Carré (25ch mode)" NameShort="ivlcarre-25ch" Comment="generated by the Open Fixture Library – https://open-fixture-library.org/minuit-une/ivl-carre" AllocateDmxChannels="25" Weight="13" Power="125" DimWidth="594.3" DimHeight="200" DimDepth="551.9">
                     <ChannelBeam Name="Control" DefaultValue="191" Highlight="255" Deflection="0" DmxByte0="1" DmxByte1="0" Constant="0" Crossfade="0" Invert="0" Precedence="LTP" ClassicPos="1">
                         <Range Name="Off" Start="0" End="23" AutoMenu="1" Centre="0"/>
                         <Range Name="Reset" Start="24" End="49" AutoMenu="1" Centre="0"/>
                         <Range Name="On" Start="50" End="255" AutoMenu="1" Centre="0"/>
@@ -60,9 +60,9 @@
                         <Range Name="0 - 360°" Start="0" End="65535" AutoMenu="1" Centre="0"/>
                     </ChannelBeam>
                     <ChannelBeam Name="Reserved" DefaultValue="0" Highlight="255" Deflection="0" DmxByte0="25" DmxByte1="0" Constant="0" Crossfade="0" Invert="0" Precedence="LTP" ClassicPos="17"/>
                 </Fixture>
-                <Fixture _CreationDate="2018-03-30#00:00:00" _ModifiedDate="2018-03-30#00:00:00" Name="M-Carré (40ch mode)" NameShort="mcarre-40ch" Comment="generated by the Open Fixture Library – https://open-fixture-library.org/minuit-une/m-carre" AllocateDmxChannels="40" Weight="13" Power="125" DimWidth="594.3" DimHeight="200" DimDepth="551.9">
+                <Fixture _CreationDate="2018-03-30#00:00:00" _ModifiedDate="2018-04-20#00:00:00" Name="IVL Carré (40ch mode)" NameShort="ivlcarre-40ch" Comment="generated by the Open Fixture Library – https://open-fixture-library.org/minuit-une/ivl-carre" AllocateDmxChannels="40" Weight="13" Power="125" DimWidth="594.3" DimHeight="200" DimDepth="551.9">
                     <ChannelBeam Name="Control" DefaultValue="191" Highlight="255" Deflection="0" DmxByte0="1" DmxByte1="0" Constant="0" Crossfade="0" Invert="0" Precedence="LTP" ClassicPos="1">
                         <Range Name="Off" Start="0" End="23" AutoMenu="1" Centre="0"/>
                         <Range Name="Reset" Start="24" End="49" AutoMenu="1" Centre="0"/>
                         <Range Name="On" Start="50" End="255" AutoMenu="1" Centre="0"/>

```
</details>

### :warning: Plugin `millumin`
*Removed files*
- minuit-une/m-carre.json

*Added files*
- minuit-une/ivl-carre.json


### :warning: Plugin `qlcplus`
*Removed files*
- Minuit-Une-M-Carré.qxf

*Added files*
- Minuit-Une-IVL-Carré.qxf

FloEdelmann commented:
<!-- GITHUB-TEST: tests/github/exports-valid.js -->
# Export files validity
(Output of test script `tests/github/exports-valid.js`.)

Test the exported files of selected fixtures against the plugins' export tests.
You can run a plugin's export tests by executing:
`$ node cli/run-export-test.js -p <plugin name> <fixtures>`

## Fixture `minuit-une/ivl-carre` changed in this PR
- d-light
  - attributes-correctness
    - :x: minuit-une/ivl-carre-25ch.xml
      <details>
      <summary>Show errors</summary>
      <ul>
      <li>Duplicate parameter name: FOCUS/TILT</li>
      <li>Duplicate parameter name: FOCUS/TILT</li>
      <li>Duplicate parameter name: FOCUS/TILT</li>
      <li>Duplicate parameter name: FOCUS/TILT</li>
      <li>Duplicate parameter name: FOCUS/PAN</li>
      </ul>
      </details>
    - :x: minuit-une/ivl-carre-40ch.xml
      <details>
      <summary>Show errors</summary>
      <ul>
      <li>Duplicate parameter name: FOCUS/TILT</li>
      <li>Duplicate parameter name: FOCUS/TILT</li>
      <li>Duplicate parameter name: FOCUS/TILT</li>
      <li>Duplicate parameter name: FOCUS/TILT</li>
      <li>Duplicate parameter name: FOCUS/PAN</li>
      <li>Duplicate parameter name: FOCUS/PAN</li>
      <li>Duplicate parameter name: FOCUS/PAN</li>
      <li>Duplicate parameter name: FOCUS/PAN</li>
      <li>Duplicate parameter name: FOCUS/PAN</li>
      </ul>
      </details>
- ecue
- millumin
- ofl
- qlcplus
  - xsd-schema-conformity
    - :white_check_mark: Minuit-Une-IVL-Carré.qxf

## Fixture `minuit-une/m-carre` changed in this PR
- d-light
  - attributes-correctness
    - :x: minuit-une/ivl-carre-25ch.xml
      <details>
      <summary>Show errors</summary>
      <ul>
      <li>Duplicate parameter name: FOCUS/TILT</li>
      <li>Duplicate parameter name: FOCUS/TILT</li>
      <li>Duplicate parameter name: FOCUS/TILT</li>
      <li>Duplicate parameter name: FOCUS/TILT</li>
      <li>Duplicate parameter name: FOCUS/PAN</li>
      </ul>
      </details>
    - :x: minuit-une/ivl-carre-40ch.xml
      <details>
      <summary>Show errors</summary>
      <ul>
      <li>Duplicate parameter name: FOCUS/TILT</li>
      <li>Duplicate parameter name: FOCUS/TILT</li>
      <li>Duplicate parameter name: FOCUS/TILT</li>
      <li>Duplicate parameter name: FOCUS/TILT</li>
      <li>Duplicate parameter name: FOCUS/PAN</li>
      <li>Duplicate parameter name: FOCUS/PAN</li>
      <li>Duplicate parameter name: FOCUS/PAN</li>
      <li>Duplicate parameter name: FOCUS/PAN</li>
      <li>Duplicate parameter name: FOCUS/PAN</li>
      </ul>
      </details>
- ecue
- millumin
- ofl
- qlcplus
  - xsd-schema-conformity
    - :white_check_mark: Minuit-Une-IVL-Carré.qxf

FloEdelmann commented:
<!-- GITHUB-TEST: tests/http-status.js -->
# Broken links
(Output of test script `tests/http-status.js`.)

There were 0 internal and 4 external links failing.

- http://www.cameolight.com/
- https://www.claypaky.it/en/home/
- http://www.adblighting.com/?page=getsupportfile&file=165
- https://www.claypaky.it/media/documents/HR_Sharpy_Manual_05.2016_EN.pdf

FloEdelmann commented:
<!-- GITHUB-TEST: tests/github/schema-version-reminder.js -->
# Schema has changed
(Output of test script `tests/github/schema-version-reminder.js`.)

With every change on the schema, its version should be incremented and tagged. See the [Fixture README](https://github.com/OpenLightingProject/open-fixture-library/blob/redirect/docs/fixture-format.md#schema) for further information.

FloEdelmann commented:
<!-- GITHUB-TEST: tests/github/export-diff.js -->
# Plugin export diff
(Output of test script `tests/github/export-diff.js`.)

You can run view your uncommited changes in plugin exports manually by executing:
`$ node cli/diff-plugin-outputs.js -p <plugin name> <fixtures>`

## Fixture `minuit-une/m-carre` modified in this PR
Fixture output to all plugins is checked.

### :warning: Plugin `d-light`
*Removed files*
- minuit-une/m-carre-25ch.xml
- minuit-une/m-carre-40ch.xml

*Added files*
- minuit-une/ivl-carre-25ch.xml
- minuit-une/ivl-carre-40ch.xml


### :warning: Plugin `ecue`
<details>
<summary><b>Changed outputted file <code>UserLibrary.xml</code></b></summary>

```diff
--- removed
+++ added
@@ -2,9 +2,9 @@
 <Document Owner="user" TypeVersion="2" SaveTimeStamp="2018-04-20#15:08:41">
     <Library>
         <Fixtures>
             <Manufacturer _CreationDate="2018-04-20#15:08:41" _ModifiedDate="2018-04-20#15:08:41" Name="Minuit Une" Comment="" Web="https://minuitune.com/">
-                <Fixture _CreationDate="2018-03-30#00:00:00" _ModifiedDate="2018-03-30#00:00:00" Name="M-Carré (25ch mode)" NameShort="mcarre-25ch" Comment="generated by the Open Fixture Library – https://open-fixture-library.org/minuit-une/m-carre" AllocateDmxChannels="25" Weight="13" Power="125" DimWidth="594.3" DimHeight="200" DimDepth="551.9">
+                <Fixture _CreationDate="2018-03-30#00:00:00" _ModifiedDate="2018-04-20#00:00:00" Name="IVL Carré (25ch mode)" NameShort="ivlcarre-25ch" Comment="generated by the Open Fixture Library – https://open-fixture-library.org/minuit-une/ivl-carre" AllocateDmxChannels="25" Weight="13" Power="125" DimWidth="594.3" DimHeight="200" DimDepth="551.9">
                     <ChannelBeam Name="Control" DefaultValue="191" Highlight="255" Deflection="0" DmxByte0="1" DmxByte1="0" Constant="0" Crossfade="0" Invert="0" Precedence="LTP" ClassicPos="1">
                         <Range Name="Off" Start="0" End="23" AutoMenu="1" Centre="0"/>
                         <Range Name="Reset" Start="24" End="49" AutoMenu="1" Centre="0"/>
                         <Range Name="On" Start="50" End="255" AutoMenu="1" Centre="0"/>
@@ -60,9 +60,9 @@
                         <Range Name="0 - 360°" Start="0" End="65535" AutoMenu="1" Centre="0"/>
                     </ChannelBeam>
                     <ChannelBeam Name="Reserved" DefaultValue="0" Highlight="255" Deflection="0" DmxByte0="25" DmxByte1="0" Constant="0" Crossfade="0" Invert="0" Precedence="LTP" ClassicPos="17"/>
                 </Fixture>
-                <Fixture _CreationDate="2018-03-30#00:00:00" _ModifiedDate="2018-03-30#00:00:00" Name="M-Carré (40ch mode)" NameShort="mcarre-40ch" Comment="generated by the Open Fixture Library – https://open-fixture-library.org/minuit-une/m-carre" AllocateDmxChannels="40" Weight="13" Power="125" DimWidth="594.3" DimHeight="200" DimDepth="551.9">
+                <Fixture _CreationDate="2018-03-30#00:00:00" _ModifiedDate="2018-04-20#00:00:00" Name="IVL Carré (40ch mode)" NameShort="ivlcarre-40ch" Comment="generated by the Open Fixture Library – https://open-fixture-library.org/minuit-une/ivl-carre" AllocateDmxChannels="40" Weight="13" Power="125" DimWidth="594.3" DimHeight="200" DimDepth="551.9">
                     <ChannelBeam Name="Control" DefaultValue="191" Highlight="255" Deflection="0" DmxByte0="1" DmxByte1="0" Constant="0" Crossfade="0" Invert="0" Precedence="LTP" ClassicPos="1">
                         <Range Name="Off" Start="0" End="23" AutoMenu="1" Centre="0"/>
                         <Range Name="Reset" Start="24" End="49" AutoMenu="1" Centre="0"/>
                         <Range Name="On" Start="50" End="255" AutoMenu="1" Centre="0"/>

```
</details>

### :warning: Plugin `millumin`
*Removed files*
- minuit-une/m-carre.json

*Added files*
- minuit-une/ivl-carre.json


### :warning: Plugin `qlcplus`
*Removed files*
- Minuit-Une-M-Carré.qxf

*Added files*
- Minuit-Une-IVL-Carré.qxf

fxedel commented on cli/make-register.js:
```diff
@@ -36,11 +36,27 @@ try {
+
+        if (fixData.$schema.endsWith(`/fixture-redirect.json`)) {
+          isRedirect = true;
+          redirectToData = JSON.parse(fs.readFileSync(path.join(fixturePath, `${fixData.redirectTo}.json`), `utf8`));
```
Is this variable used anywhere?

fxedel commented on cli/make-register.js:
```diff
@@ -112,7 +132,9 @@ const lastActionRankMapping = {
 };
 
 // add fixture list sorted by lastModifyDate
-register.lastUpdated = Object.keys(register.filesystem).sort((a, b) => {
```
Maybe ``!(`redirectTo` in register.filesystem[fixtureKey])`` would be clearer?

FloEdelmann replied:
But we actually use `lastModifyDate` and `lastAction` inside the following sort function, so filtering for those (assuming they only come in pairs) makes more sense IMO.

fxedel commented on lib/create-github-pr.js:
```diff
@@ -328,7 +328,9 @@ module.exports = function createPullRequest(out, callback) {
     // add fixture list sorted by lastModifyDate
-    newRegister.lastUpdated = Object.keys(newRegister.filesystem).sort((a, b) => {
+    newRegister.lastUpdated = Object.keys(newRegister.filesystem).filter(
+      fixtureKey => `lastModifyDate` in register.filesystem[fixtureKey]
```
Same here

fxedel commented on tests/fixture-valid.js:
```diff
@@ -45,6 +46,23 @@ module.exports = function checkFixture(manKey, fixKey, fixtureJson, uniqueValues
+    if (!valid) {
+      result.errors.push(module.exports.getErrorString(`File does not match schema.`, validate.errors));
+    }
+
```
Please forbid redirects as redirect targets (e. g. to avoid circular redirects)

FloEdelmann commented:
<!-- GITHUB-TEST: tests/github/export-diff.js -->
# Plugin export diff
(Output of test script `tests/github/export-diff.js`.)

You can run view your uncommited changes in plugin exports manually by executing:
`$ node cli/diff-plugin-outputs.js -p <plugin name> <fixtures>`

## Fixture `minuit-une/m-carre` modified in this PR
Fixture output to all plugins is checked.

### :warning: Plugin `d-light`
*Removed files*
- minuit-une/m-carre-25ch.xml
- minuit-une/m-carre-40ch.xml

*Added files*
- minuit-une/ivl-carre-25ch.xml
- minuit-une/ivl-carre-40ch.xml


### :warning: Plugin `ecue`
<details>
<summary><b>Changed outputted file <code>UserLibrary.xml</code></b></summary>

```diff
--- removed
+++ added
@@ -2,9 +2,9 @@
 <Document Owner="user" TypeVersion="2" SaveTimeStamp="2018-04-21#15:03:42">
     <Library>
         <Fixtures>
             <Manufacturer _CreationDate="2018-04-21#15:03:42" _ModifiedDate="2018-04-21#15:03:42" Name="Minuit Une" Comment="" Web="https://minuitune.com/">
-                <Fixture _CreationDate="2018-03-30#00:00:00" _ModifiedDate="2018-03-30#00:00:00" Name="M-Carré (25ch mode)" NameShort="mcarre-25ch" Comment="generated by the Open Fixture Library – https://open-fixture-library.org/minuit-une/m-carre" AllocateDmxChannels="25" Weight="13" Power="125" DimWidth="594.3" DimHeight="200" DimDepth="551.9">
+                <Fixture _CreationDate="2018-03-30#00:00:00" _ModifiedDate="2018-04-20#00:00:00" Name="IVL Carré (25ch mode)" NameShort="ivlcarre-25ch" Comment="generated by the Open Fixture Library – https://open-fixture-library.org/minuit-une/ivl-carre" AllocateDmxChannels="25" Weight="13" Power="125" DimWidth="594.3" DimHeight="200" DimDepth="551.9">
                     <ChannelBeam Name="Control" DefaultValue="191" Highlight="255" Deflection="0" DmxByte0="1" DmxByte1="0" Constant="0" Crossfade="0" Invert="0" Precedence="LTP" ClassicPos="1">
                         <Range Name="Off" Start="0" End="23" AutoMenu="1" Centre="0"/>
                         <Range Name="Reset" Start="24" End="49" AutoMenu="1" Centre="0"/>
                         <Range Name="On" Start="50" End="255" AutoMenu="1" Centre="0"/>
@@ -60,9 +60,9 @@
                         <Range Name="0 - 360°" Start="0" End="65535" AutoMenu="1" Centre="0"/>
                     </ChannelBeam>
                     <ChannelBeam Name="Reserved" DefaultValue="0" Highlight="255" Deflection="0" DmxByte0="25" DmxByte1="0" Constant="0" Crossfade="0" Invert="0" Precedence="LTP" ClassicPos="17"/>
                 </Fixture>
-                <Fixture _CreationDate="2018-03-30#00:00:00" _ModifiedDate="2018-03-30#00:00:00" Name="M-Carré (40ch mode)" NameShort="mcarre-40ch" Comment="generated by the Open Fixture Library – https://open-fixture-library.org/minuit-une/m-carre" AllocateDmxChannels="40" Weight="13" Power="125" DimWidth="594.3" DimHeight="200" DimDepth="551.9">
+                <Fixture _CreationDate="2018-03-30#00:00:00" _ModifiedDate="2018-04-20#00:00:00" Name="IVL Carré (40ch mode)" NameShort="ivlcarre-40ch" Comment="generated by the Open Fixture Library – https://open-fixture-library.org/minuit-une/ivl-carre" AllocateDmxChannels="40" Weight="13" Power="125" DimWidth="594.3" DimHeight="200" DimDepth="551.9">
                     <ChannelBeam Name="Control" DefaultValue="191" Highlight="255" Deflection="0" DmxByte0="1" DmxByte1="0" Constant="0" Crossfade="0" Invert="0" Precedence="LTP" ClassicPos="1">
                         <Range Name="Off" Start="0" End="23" AutoMenu="1" Centre="0"/>
                         <Range Name="Reset" Start="24" End="49" AutoMenu="1" Centre="0"/>
                         <Range Name="On" Start="50" End="255" AutoMenu="1" Centre="0"/>

```
</details>

### :warning: Plugin `millumin`
*Removed files*
- minuit-une/m-carre.json

*Added files*
- minuit-une/ivl-carre.json


### :warning: Plugin `qlcplus`
*Removed files*
- Minuit-Une-M-Carré.qxf

*Added files*
- Minuit-Une-IVL-Carré.qxf

FloEdelmann commented:
<!-- GITHUB-TEST: tests/github/export-diff.js -->
# Plugin export diff
(Output of test script `tests/github/export-diff.js`.)

You can run view your uncommited changes in plugin exports manually by executing:
`$ node cli/diff-plugin-outputs.js -p <plugin name> <fixtures>`

## Fixture `minuit-une/m-carre` modified in this PR
Fixture output to all plugins is checked.

### :warning: Plugin `d-light`
*Removed files*
- minuit-une/m-carre-25ch.xml
- minuit-une/m-carre-40ch.xml

*Added files*
- minuit-une/ivl-carre-25ch.xml
- minuit-une/ivl-carre-40ch.xml


### :warning: Plugin `ecue`
<details>
<summary><b>Changed outputted file <code>UserLibrary.xml</code></b></summary>

```diff
--- removed
+++ added
@@ -2,9 +2,9 @@
 <Document Owner="user" TypeVersion="2" SaveTimeStamp="2018-04-21#15:31:21">
     <Library>
         <Fixtures>
             <Manufacturer _CreationDate="2018-04-21#15:31:21" _ModifiedDate="2018-04-21#15:31:21" Name="Minuit Une" Comment="" Web="https://minuitune.com/">
-                <Fixture _CreationDate="2018-03-30#00:00:00" _ModifiedDate="2018-03-30#00:00:00" Name="M-Carré (25ch mode)" NameShort="mcarre-25ch" Comment="generated by the Open Fixture Library – https://open-fixture-library.org/minuit-une/m-carre" AllocateDmxChannels="25" Weight="13" Power="125" DimWidth="594.3" DimHeight="200" DimDepth="551.9">
+                <Fixture _CreationDate="2018-03-30#00:00:00" _ModifiedDate="2018-04-20#00:00:00" Name="IVL Carré (25ch mode)" NameShort="ivlcarre-25ch" Comment="generated by the Open Fixture Library – https://open-fixture-library.org/minuit-une/ivl-carre" AllocateDmxChannels="25" Weight="13" Power="125" DimWidth="594.3" DimHeight="200" DimDepth="551.9">
                     <ChannelBeam Name="Control" DefaultValue="191" Highlight="255" Deflection="0" DmxByte0="1" DmxByte1="0" Constant="0" Crossfade="0" Invert="0" Precedence="LTP" ClassicPos="1">
                         <Range Name="Off" Start="0" End="23" AutoMenu="1" Centre="0"/>
                         <Range Name="Reset" Start="24" End="49" AutoMenu="1" Centre="0"/>
                         <Range Name="On" Start="50" End="255" AutoMenu="1" Centre="0"/>
@@ -60,9 +60,9 @@
                         <Range Name="0 - 360°" Start="0" End="65535" AutoMenu="1" Centre="0"/>
                     </ChannelBeam>
                     <ChannelBeam Name="Reserved" DefaultValue="0" Highlight="255" Deflection="0" DmxByte0="25" DmxByte1="0" Constant="0" Crossfade="0" Invert="0" Precedence="LTP" ClassicPos="17"/>
                 </Fixture>
-                <Fixture _CreationDate="2018-03-30#00:00:00" _ModifiedDate="2018-03-30#00:00:00" Name="M-Carré (40ch mode)" NameShort="mcarre-40ch" Comment="generated by the Open Fixture Library – https://open-fixture-library.org/minuit-une/m-carre" AllocateDmxChannels="40" Weight="13" Power="125" DimWidth="594.3" DimHeight="200" DimDepth="551.9">
+                <Fixture _CreationDate="2018-03-30#00:00:00" _ModifiedDate="2018-04-20#00:00:00" Name="IVL Carré (40ch mode)" NameShort="ivlcarre-40ch" Comment="generated by the Open Fixture Library – https://open-fixture-library.org/minuit-une/ivl-carre" AllocateDmxChannels="40" Weight="13" Power="125" DimWidth="594.3" DimHeight="200" DimDepth="551.9">
                     <ChannelBeam Name="Control" DefaultValue="191" Highlight="255" Deflection="0" DmxByte0="1" DmxByte1="0" Constant="0" Crossfade="0" Invert="0" Precedence="LTP" ClassicPos="1">
                         <Range Name="Off" Start="0" End="23" AutoMenu="1" Centre="0"/>
                         <Range Name="Reset" Start="24" End="49" AutoMenu="1" Centre="0"/>
                         <Range Name="On" Start="50" End="255" AutoMenu="1" Centre="0"/>

```
</details>

### :warning: Plugin `millumin`
*Removed files*
- minuit-une/m-carre.json

*Added files*
- minuit-une/ivl-carre.json


### :warning: Plugin `qlcplus`
*Removed files*
- Minuit-Une-M-Carré.qxf

*Added files*
- Minuit-Une-IVL-Carré.qxf
